### DISCUSSION TITLE
⏫🆙 TshetUinh.js v0.15 更動概要及遷移指南

### DISCUSSION THREAD

**Discussion by syimyuzya** [nk2028/qieyun-js#43] [Announcements]
TshetUinh.js v0.15 已於 2024-09-02 發佈！

推導器 v0.2 目前 beta 測試中！

- 試用新版推導器：https://syimyuzya.github.io/qieyun-autoderiver/
- 試用新版 TshetUinh.js：可直接在新版推導器的控制台試用（請暫用原名 `Qieyun`），或者 `npm i tshet-uinh` 在本機試用

本文列出遷移時需要注意的更動。

## 音韻地位體系

v0.15 音韻地位恢復了單一體系及較嚴格的 `音韻地位.驗證`。該體系（及內置資料）改進自之前 v0.13 的 v2 體系。

[切韻新韻圖](https://phesoca.com/rime-table/)已在使用該體系，可前往瀏覽。

以下為新舊版差異及遷移時需要注意的點：

（注意：本文僅列出與舊版的差異點，完整的音韻地位規則請參見 `音韻地位.驗證` 的文檔）

音韻地位表達形式與 v0.13 的 v2 體系不同者：

- 用字更動：
  - 眞 → 真（與字頭一致）
    - 例：「真」章開三**真**平
  - 欣 → 殷（用切韻韻目）
    - 例：「欣」曉開三C**殷**平
- 「呼」更動：
  - 虞韻：開合中立 → 合口（與魚韻開合對應）
    - 例：「朱」章**合**三虞平
  - 侯韻：開口 → 開合中立（與尤、東韻平行）
    - 例：「樓」來一侯平
  - 幽韻：開合中立 → 開口（與侵韻平行）
    - 例：「糾」見**開**三A幽上
- 「重紐」改為「類」，`音韻地位.判斷` 表達式中格式形如 `B類`、`C類`、`AB類`、`不分類` 等
  - 所有三等鈍聲母均標明「類」（銳聲母或非三等則均不標「類」）
    - 例：「方」幫三**C**陽平、「將」精開三陽平
  - 云母均標明 B/C 類
    - 例：「為」云合三**B**支平、「王」云合三**C**陽平
  - 庚韻原不標重紐，現鈍聲母均標明 B 類
    - 例：「明」明三**B**庚平
  - 清韻原不標重紐，現鈍聲母均標明 A 類
    - 例：「名」明三**A**清平
  - 幽、蒸韻原本除開口牙喉B類外不標重紐，現幽韻均標明 A/B 類、蒸韻均標明 B/C 類
    - 例：「幽」影**開**三**A**幽平、「冰」幫三**B**蒸平、「域」云合三**B**蒸入
  - 「輕脣」現可用「幫組 C類」直接判斷，不需再羅列輕脣韻
- 「等」更動：
  - 三等韻在拼端組時作為「四等」（如地、爹、丟等），一些判斷「三等」「四等」的條件需要注意端組（比如可用 `TshetUinh.表達式.四等韻` 判斷或排除僅四等的韻）
    - 例：「地」定開**四**脂去、「爹」端開**四**麻平
  - 但二等韻拼端組仍作為「二等」
    - 例：「打」端開二庚上、「箉」定開二佳上
- 其他更動：
  - 元韻由臻攝改歸**山**攝

v0.13 的 v2 中存在，新版取消了的情形：

- 匣母三等

新版仍然存在，可能需要注意的邊緣地位：

- 祭廢韻平上聲（搭配章組、云以日母，《廣韻》寄於齊、灰、咍韻
  - 例：「茝」昌開三廢上
- 蒸韻一般脣音及合口為 B 類，開口牙喉音為 C 類，但亦存在少量開口牙喉音 B 類
  - 例：「憶」影開C蒸入、「抑」影開B蒸入
- 幽韻一般脣音為 B 類，牙喉音為 A 類，但亦存在少量牙喉音 B 類
- 陽韻存在A類，資料僅「𩦠」小韻一筆
  - 例：並三A陽上

v0.13 的 v2ext 額外允許的組合，新版亦全部取消（列舉於此僅供備考）：

- 莊組眞欣韻
- 寒歌韻脣音「呼」標為「開」
- 咍嚴韻脣音、凡韻非脣音
- 清韻B類
- 除已知地位外，其他端組拼二三等韻
- 知組拼一四等韻
- 章組云以日母拼齊灰咍韻
- 精組二等、莊組一四等、章組以日母非三等
- 云母非三等

PS 為什麼新版的體系沒起名字呢？比如叫 v3？

答：因為之前是為了讓多個體系並存於同一版本所以有名字。但如今既然發現[這種設計有害](https://github.com/nk2028/tshet-uinh-js/discussions/37)並取消掉了，整個版本就只對應一個體系，那直接說「v0.15 的體系」就可以了。

## TshetUinh.js API 更動

- 移除 `Qieyun.分析體系` 模塊，不再支持相關功能
- 移除 `Qieyun.推導方案` 模塊，其已被 [tshet-uinh-deriver-tools](https://github.com/nk2028/tshet-uinh-deriver-tools) 取代
- 移除 `音韻地位.編碼` 及 `.from編碼`，相關功能獨立為 `TshetUinh.壓縮表示` 模塊
  - （※ 注意：新舊版編碼**不相容**，不可混用）
- `音韻地位.屬於` 移除以下表達式：
  - 「輕脣韻」（可用「脣音 C類」代之，該條件範圍更準確）
  - 「次入韻」（可用「去聲 (蟹攝 非 佳韻 或 脂微韻)」代之）
- `音韻地位.最簡描述` 改為 `.簡略描述`（注意：其簡省規則尚未敲定，請不要用簡略描述進行文本比較）
- `音韻地位.from描述` 預設僅認別完整描述，需指定額外參數以令其識別簡略描述
- 檢索結果物件亦包含 `字頭`（按字頭和按音韻地位查詢均會得到相同格式結果），此外其 `解釋` 改稱 `釋義`， `韻部原貌` 字串改為 `來源` 物件，其中亦包含來源文獻名等
- `音韻地位` 增加 `.字母`（三十六字母）、`.韻圖等` 功能
- `TshetUinh.資料` 增加 `.廣韻` 子模塊，可依小韻檢索內置廣韻資料（包括四個沒有音韻地位的訛字訛切小韻）

此外，TshetUinh.js 的工具鏈亦更新，構建及運行均需要 Node 18 以上。

## 遷移要點

注意：舊版推導方案代碼**不能**運行於 TshetUinh.js v0.15，需要更新代碼後方可使用。

（舊版推導器之後仍將繼續在線，可用於運行舊方案，但不會收到功能或介面更新）

更新方案時的注意點：

- 「類」與之前的「重紐」相比，改進了覆蓋範圍：
  - 鈍聲母三等均有「類」，銳聲母或非三等則均不標「類」
  - 「A類」除之前的「重紐A類」外，還包含同為A類的清韻、麻韻、幽韻（部分）
  - 「B類」除之前的「重紐B類」外，還包含同為B類的云母（可搭配B類的韻）、庚韻、蒸幽韻（部分）
  - 「C類」與之前的「鈍音 三等 不分重紐」相比：不含實為A或B類的庚清幽麻韻，不含部分云母（即上述B類部分）
  - 「不分類」等同於「非 (鈍音 三等)」或者「銳音 或 非 三等」
- 判斷「呼」時，請注意在新版本：
  - 「合口」會包含虞韻
  - 「開口」會包含幽韻，不含侯韻
  - 「開合中立」會包含侯韻，不含幽、虞韻
- 三等韻拼端組時是「四等」，如需與四等韻區分，可考慮使用 `TshetUinh.表達式.四等韻` 判斷

## 推導方案相關新功能

因為是「遷移」指南所以此處不贅述。可參見推導器內置說明（點擊介面頂部「？」圖示）。

**Comment by guavajuice** [comment]
庚韻原不標重紐，現鈍聲母均標明 B 類
例：「明」明三B庚平
清韻原不標重紐，現鈍聲母均標明 A 類
例：「名」明三A清平
--------------------------------------------
我覺得這箇操作不妥。實際上，清韻自身就是一箇重紐韻，有「碧」、「辟」對立。

**Reply by syimyuzya** [reply]
Qieyun.js 曾經在 v0.8-v0.12 期間（2021-2022）確實採用過閣下所說的 poem《廣韻字音表》的音韻地位體系。但據我們所知，清、庚韻為對應的 A、B 類，以及《廣韻》增補小韻「碧」其音實屬庚（陌）韻，已臻於共識。《字音表》的體系才是不合適的一方。故 Qieyun.js 早在兩年前（2022）的 v0.13新體系就已經修正了這點（但暫時兼容舊分析）。而如今 v0.15 做的則是移除這個暫時的兼容，並藉此更清晰地呈現音韻地位，以方便將來的工作。

PS 從所提的幾點來看，閣下可能尚未了解過近年的一些進展。如果想得知更多前沿信息的話，知乎 Sliark 有相關的音韻學討論 QQ 群，可前去詢問如何加入。

**Reply by syimyuzya** [reply]
已經說過了，這超出了本主題的範疇。討論更多理據請加入（或建立）討論組進行。

**Reply by guavajuice** [reply]
我剛剛建了一箇群 623996198

**Comment by guavajuice** [comment]
三等韻在拼端組時作為「四等」（如地、爹、丟等），一些判斷「三等」「四等」的條件需要注意端組（比如可用 Qieyun.表達式.四等韻 判斷或排除僅四等的韻）
例：「地」定開四脂去、「爹」端開四麻平
------------------------------------------------
「地」可以看成是重紐三等。

**Reply by syimyuzya** [reply]
「三等」和「非三等」在切韻時代是最需要區分的特徵，不可混。而且它是整個音節上的（且偏向音節前半的）特徵，並不像舊觀點認為的扔給韻頭就行。（也和上一條一樣，這點在知乎等處已討論並發展了若干年，相關項目諸如新韻圖也已經提前將其應用了兩年）如今 v0.15 大更新，正是為了及時跟進這兩年來的成果，以方便未來的更多工作。

**Reply by syimyuzya** [reply]
【或者更通俗些類比：這就有點類似希頂語 yε、4ε 這兩個音節（或者對比 qε、kε 也可以），韻母一樣都是 [i]，但看音節特點，前者是齶化的，後者不是，並不是同一類音節

**Comment by guavajuice** [comment]
祭廢韻平上聲（搭配章組、云以日母，《廣韻》寄於齊、灰、咍韻
例：「茝」昌開三廢上
-------------------------------------------------------------------------
這箇操作亦是不妥的。祭部有特殊的結構乃是-l，只有去聲。
直接拼合即可！

**Reply by syimyuzya** [reply]
這個音節尾特點並不與常說的「次入韻」對應。諸如齊、脂等韻去聲，中古在一段時期內也與祭韻一樣特殊，且不分上古來源一律如此（可參見 unt 的《齊梁陳隋押韻材料的數理分析》重製版）。故「次入韻」實無此限制。

**Reply by syimyuzya** [reply]
在此之上的更多討論，則已超出 Qieyun.js 的範疇。如果想了解更多，知乎 Sliark 有相關的音韻學討論 QQ 群，可前往參與。

**Reply by syimyuzya** [reply]
群號 739872788
群裡目前相當活躍，是很適合參與的地方，歡迎加入～

**Comment by guavajuice** [comment]
又發現一箇問題“礥”應該是“下珎反”，怎麼變成了“下殄切”？
如果沒有考古上的證據，就不要亂改。

**Reply by syimyuzya** [reply]
這也是兩年多前早考察過的問題了（

【PS 閣下提過的四件事不僅無一與「更動概要及遷移指南」相關，且無一例外全是跨度兩年以上的挖墳話題。要是接連不斷的話那可就是 spam 了

**Reply by guavajuice** [reply]
那請給出「考古」方面的證據。如果用一種不正確的擬音，是不能說明問題的。

**Reply by guavajuice** [reply]
> 2022.9.5 补：“下珍反”当为“下殄切”之讹。《集韵》“礥”新增有上声 ghenq 音，《[篆隶万象名义]》有“臔，下殄反”（ghenq）说明“下殄”是事实存在的切语。另外《新撰字镜》有“礥，下弁反”（去声 ghenh）也是一个旁证。

剛看到，這無法說明「礥」跟「臔」是一箇字。

**Reply by syimyuzya** [reply]
好了，早就三次離題告知了。
